### ISSUE TITLE
make HAL provenance-friendly

### ISSUE THREAD

**Issue by hawkw** [hawkw/mycelium#366]
the Mycelium HAL's address types (and other code that does on MMIO)  was mostly written before Rust had ways to do int2ptr conversions without totally violating pointer provenance. we should update this code to be less egregiously wrong.

this basically means replacing all the int2ptr casts with `ptr::from_exposed_addr` and friends.

**Comment by hawkw** [comment]
we might also be better off changing the internal representation of `VAddr` from `VAddr(usize)` to `VAddr(*const ())` so that when you construct one from a ptr it's "more provenance-preserving", and make the `VAddr::from_usize` constrictor be `ptr::from_exposed_addr`...

**Comment by hawkw** [comment]
@myrrlyn do you wanna mess with this?

**Comment by myrrlyn** [comment]
sure. fwiw the standard library seems to Love storing `*mut ()` instead of `*const ()` for some reason so if `VAddr` occasionally allows write permissions it might be worth using that as the storage. idk i'm not a compilers boy. in The Future tm once `funty` gets cool (and able to provide `const fn`s...) this won't matter, it'll store the Correct pointer always.

anyway yeah i'll poke

**Comment by hawkw** [comment]
yeah, it should probably be `*mut ()` since it will grant write permissions through unsafe code --- that was what i meant.

**Comment by hawkw** [comment]
@myrrlyn is this still something you're interested in taking a crack at? i might do it at some point, otherwise.

**Comment by myrrlyn** [comment]
yes but feel free to beat me to it

**Comment by hawkw** [comment]
i'm probably not going to start any time soon, so, i probably won't. but i figured i'd check.
